File: src_fortran/sub/matrmult.f
c
c
c
c
c
c
      subroutine MATRMULT(A,m,p,B,p1,n,C,m1,n1)
c     Urs Kradolfer, 28.3.1987
      implicit none
c     input:  (mxp)-matrix A
c             (p1xn)-matrix B
c     output: (m1xn1)-matrix C = A*B
      integer m,p, p1,n, m1,n1 ,
     &        i,j,k
      real A(m,p), B(p1,n), C(m1,n1) ,
     &     s
      if(m.ne.m1)stop'Matrices cannot be multiplied !'
      if(p.ne.p1)stop'Matrices cannot be multiplied !'
      if(n.ne.n1)stop'Matrices cannot be multiplied !'
      do i=1,m
         do j=1,n
            s=0.
            do k=1,p
               s=s+A(i,k)*B(k,j)
            enddo
            C(i,j)=s
         enddo
      enddo
      return
      end ! of subr. matrmult

File: src/sub/matrmult.c
#include <stddef.h>

void matrmult(const double *a, int m, int p, const double *b, int p1, int n,
              double *c, int m1, int n1) {
    int i;
    int j;
    int k;

    if (a == NULL || b == NULL || c == NULL || m <= 0 || p <= 0 || n <= 0 ||
        p1 <= 0 || m1 <= 0 || n1 <= 0) {
        return;
    }

    if (m != m1 || p != p1 || n != n1) {
        return;
    }

    for (i = 0; i < m; ++i) {
        for (j = 0; j < n; ++j) {
            double s = 0.0;
            for (k = 0; k < p; ++k) {
                s += a[i * p + k] * b[k * n + j];
            }
            c[i * n + j] = s;
        }
    }
}
